### PR TITLE
Add dangling comment handling for `lambda` expressions

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

This PR adds dangling comment handling for `lambda` expressions. In short, comments around the `lambda` and the `:` are all considered dangling. Comments that come between the `lambda` and the `:` may be moved after the colon for simplicity (this is an odd position for a comment anyway), unless they also precede the lambda parameters, in which case they're formatted before the parameters.

Closes https://github.com/astral-sh/ruff/issues/7470.

## Test Plan

`cargo test`

No change in similarity.

Before:

| project      | similarity index  | total files       | changed files     |
|--------------|------------------:|------------------:|------------------:|
| cpython      |           0.76083 |              1789 |              1632 |
| django       |           0.99982 |              2760 |                37 |
| transformers |           0.99957 |              2587 |               398 |
| twine        |           1.00000 |                33 |                 0 |
| typeshed     |           0.99983 |              3496 |                18 |
| warehouse    |           0.99929 |               648 |                16 |
| zulip        |           0.99962 |              1437 |                22 |

After:

| project      | similarity index  | total files       | changed files     |
|--------------|------------------:|------------------:|------------------:|
| cpython      |           0.76083 |              1789 |              1632 |
| django       |           0.99982 |              2760 |                37 |
| transformers |           0.99957 |              2587 |               398 |
| twine        |           1.00000 |                33 |                 0 |
| typeshed     |           0.99983 |              3496 |                18 |
| warehouse    |           0.99929 |               648 |                16 |
| zulip        |           0.99962 |              1437 |                22 |

### CODE REVIEW

File: crates/ruff_python_formatter/src/comments/placement.rs [hunk 1]
```diff
@@ -1623,6 +1624,119 @@ fn handle_named_expr_comment<'a>(
...
+/// Handles comments around the `:` token in a lambda expression.
+///
+/// For parameterized lambdas, both the comments between the `lambda` and the parameters, and the
+/// comments between the parameters and the body, are considered dangling, as is the case for all
+/// of the following:
+///
+/// ```python
+/// (
+///     lambda  # 1
+///     # 2
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/comments/placement.rs:1636]: What's the reason for making all comments dangling vs forcing line breaks when formatting the lambda if the argument e.g. has leading comments?

File: crates/ruff_python_formatter/src/comments/placement.rs [hunk 2]
```diff
@@ -1623,6 +1624,119 @@ fn handle_named_expr_comment<'a>(
...
+    comment: DecoratedComment<'a>,
+    lambda: &'a ast::ExprLambda,
+    locator: &Locator,
+) -> CommentPlacement<'a> {
+    if let Some(parameters) = lambda.parameters.as_deref() {
+        // Comments between the `lambda` and the parameters are dangling on the lambda:
+        // ```python
+        // (
+        //     lambda  # comment
+        //     x:
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/comments/placement.rs:1666]: Can we consider own line comments as leading comments?

File: crates/ruff_python_formatter/src/comments/placement.rs [hunk 3]
```diff
@@ -1623,6 +1624,119 @@ fn handle_named_expr_comment<'a>(
...
+        // ```
+        if comment.start() < parameters.start() {
+            return CommentPlacement::dangling(comment.enclosing_node(), comment);
+        }
+
+        // Comments between the parameters and the body are dangling on the lambda:
+        // ```python
+        // (
+        //     lambda x:  # comment
+        //     y
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/comments/placement.rs:1678]: Same here, what about own line comments? Could they be leading comments of the parameter

File: crates/ruff_python_formatter/src/comments/placement.rs (hunk 2 above)
Comment (charliermarsh) [reply crates/ruff_python_formatter/src/comments/placement.rs:1666]: How would that work for this case?

```python
(
  lambda  # 1
  # 2
  : # 3
  # 4
  x
)
```

We could make `# 4` a leading comment, but I don't think any of the others can be treated as such.

File: crates/ruff_python_formatter/src/comments/placement.rs (hunk 1 above)
Comment (charliermarsh) [reply crates/ruff_python_formatter/src/comments/placement.rs:1636]: Are you suggesting that `# 1` here should be a leading comment? Would you expect it to be moved to an own-line comment, or rendered as-is in this snippet?

File: crates/ruff_python_formatter/src/comments/placement.rs (hunk 3 above)
Comment (charliermarsh) [reply crates/ruff_python_formatter/src/comments/placement.rs:1678]: Yes the own-line comments could in this case. I found it simpler to have fewer cases but I can change that.

File: crates/ruff_python_formatter/src/comments/placement.rs (hunk 1 above)
Comment (konstin) [reply crates/ruff_python_formatter/src/comments/placement.rs:1636]: we could do it similar to lists, where 1 is dangling and 2 is leading:
```python
x = [  # 1
    # 2
    "a",
    "b",
    "c",
]
```

```
7..10, None, Some((ExprConstant, 23..26)), (ExprList, 4..47), "# 1"
15..18, None, Some((ExprConstant, 23..26)), (ExprList, 4..47), "# 2"
{
    Node {
        kind: ExprList,
        range: 4..47,
        source: `[  # 1⏎`,
    }: {
        "leading": [],
        "dangling": [
            SourceComment {
                text: "# 1",
                position: EndOfLine,
                formatted: true,
            },
        ],
        "trailing": [],
    },
    Node {
        kind: ExprConstant,
        range: 23..26,
        source: `"a"`,
    }: {
        "leading": [
            SourceComment {
                text: "# 2",
                position: OwnLine,
                formatted: true,
            },
        ],
        "dangling": [],
        "trailing": [],
    },
}
x = [  # 1
    # 2
    "a",
    "b",
    "c",
]
```

**Review by konstin — APPROVED** [review]

**Comment by charliermarsh** [comment]
Okay, I played with this one too and I think the current solution is simpler than trying to make some of these leading comments. Part of the issue is that if we make any of these leading or trailing comments, it's hard to disambiguate between these cases:

```python
(
    lambda:
    # comment
    x
)

(
    lambda:  (
        # comment
        x
    )
)
```

It's not _impossible_, but it requires more special-casing in placement, and then (more importantly) the `lambda` formatter becomes more complicated, since the logic for inserting a hard line break after the `:` has to take into account whether the body has leading own-line comments within or outside of parentheses.

**Pull request merged**